Reimplement the Issue Labeler using GitHub Workflows

jeffhandley opened the pull request:
This is a reimplementation of the Issue Labeler that uses GitHub workflows instead of requiring GitHub app services and other resources. See [issue-labeler/README.md (jeffhandley/github-workflows)](https://github.com/jeffhandley/issue-labeler/blob/jeffhandley/github-workflows/README.md) for information about its components.

At a high level:
* GitHub issue and pull request data is downloaded from a workflow and stored into the GitHub Action Cache
* That data is processed through ML.NET, and the trained model is also stored into the GitHub Action Cache
* Models can be stored in a 'staging' slot in the cache, and promoted to 'LIVE' for prediction use
* A lightweight Predictor app is precompiled and stored into the GitHub Action Cache
* Workflows run when issues and pull requests are opened, and the Predictor app is used to predict the labels for the issue or pull request

Some new capabilities this approach offers:
* Repositories can onboard themselves by replicating just 4 workflows into the .github/workflows folder (See https://github.com/jeffhandley/issue-labeler-test/tree/main/.github/workflows for an example)
* Models can be trained using data from another repository, which we have needed occasionally
* Models can be retrained at any time by running the train and promote workflows
* Output from the Predictor app is visible on the Action runs, showing the top 3 predicted labels and their scores

With the precompiled Predictor app, it typically takes about 3 seconds for the prediction workflow to execute (plus the short delay from the workflow getting triggered). The training process runs for dotnet/runtime's data in about 2 hours (compared to the ~6-hour process of the previous implementation). Most of that time savings is because we no longer use AutoML to build the model. The previous implementation was using AutoML but always forcing a single approach to be used anyway; the new implementation matches the end result of the prior implementation.

The implementation around saving the data files and building the models is simplified as well. The previous implementation had provisions for lots of additonal data to be in the TSV files that didn't actually get used. There was also a lot of legacy code around data moving across repos, such as when dotnet/corefx moved into dotnet/runtime--none of that logic was necessary anymore.

RussKie commented on src/Common/DataFileUtils.cs:
```diff
@@ -0,0 +1,46 @@
+        .Replace('\r', ' ')
+        .Replace('\n', ' ')
+        .Replace('\t', ' ')
+        .Replace('"', '`');
```
While working on an unrelated project (which involved pulling data from GitHub) I had a similar need to sanitise the content. I found that there are more things worth removing from the content before converting it into a model. E.g., html tags, markdown tables, etc.:
* https://github.com/RussKie/GitHubDataScrubber/blob/55dfa4e71cf11b19c43073fd1861f117a9300370/Program.cs#L143-L154
* https://github.com/RussKie/GitHubDataScrubber/blob/55dfa4e71cf11b19c43073fd1861f117a9300370/Program.cs#L174-L191

Whilst in my project I needed to keep only "human conversational" content (e.g., no code snippets), and here we want to preserve most of the content (including code), I feel that there are still similarities - e.g., I don't think it's worth while keeping markdown tables as those are usually used for auxiliary information (such as perf data which doesn't affect the comprehension of the content and, thus, doesn't affect the labelling).

RussKie replied:
It is also worth considering scrubbing any PII or PII-alike content (such as @<!---->github aliases and URLs).

jeffhandley replied:
Any time we post performance test results, those will be in markdown tables, and that can have data that reveals the classes. This implementation aimed to keep parity with the data going into the model, and I'm inclined to keep this as simple as possible, letting the model do the work rather than trying to build an additional layer of heuristics.

jeffhandley replied:
> It is also worth considering scrubbing any PII or PII-alike content (such as @github aliases and URLs).

Anything posted is already public, so I'm not sure there's value in trying to strip this out. GitHub aliases and URLs could potentially feed the data too.

RussKie commented on src/Downloader/Args.cs:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    public static (
+        string org,
+        string[] repos,
+        string githubToken,
+        string? issuesPath,
+        int? issueLimit,
+        string? pullsPath,
+        int? pullLimit,
+        int? pageSize,
+        int? pageLimit,
+        int[] retries,
+        Predicate<string> labelPredicate,
+        bool verbose
```
Is it worth converting this into a struct or a record? To simplify the code and make it more extensibility-friendly (e.g., adding a new property will be a simple task vs creating new variables all over the place).

jeffhandley replied:
Yeah, that'd be all right. I was trying to keep it as simple as possible, but your illustration of moving to structs looks good.

jeffhandley commented on src/GitHubClient/GitHubApi.cs:
```diff
@@ -0,0 +1,314 @@
+
+    public static async Task<string?> AddLabel(string githubToken, string org, string repo, string type, ulong number, string label)
+    {
+        using var client = CreateRestClient(githubToken);
```
Good catch. We should presumably just create and cache a single static REST client and a single GraphQL client in the app. With that, it'd be fine to still just construct a HttpClient within each of those since that'd be at most 2 HttpClients. Or we could create a single static HttpClient and reuse it across both of the GitHub clients.

RussKie commented on src/Downloader/Args.cs:
```diff
@@ -0,0 +1,226 @@
+    public int? PageLimit { get; set; }
+    public int[] Retries { get; set; }
+    public Predicate<string> LabelPredicate { get; set; }
+    public bool Verbose { get; set; }
```
I was thinking of maybe trying to unify the args structs, most of them dup quite a bit.

jeffhandley replied:
I think we've squeezed as much out of the arg factoring as is worthwhile. Once the next version of System.CommandLine is published publicly, I'll likely want to just adopt that into here next time we touch it.

jeffhandley commented:
@RussKie I did more testing and decided to make another round of changes to the arg parsing.

1. Changed the names from 'Config' to 'Args' and collapsed from 2 files into 1, putting the parse functions into the structs as statics (instead of having to bounce between the 2 files)
2. I found the issue-data and pull-data default value handling wasn't working as I thought it was. Those defaults didn't allow choosing to only process one or the other and it applied the default to make it always run both. We need to be able to run them independently.
3. I matched up a couple of arg struct property names that weren't consistent (IssuesPath vs. IssueDataPath)
4. I refactored out some arg helpers to reuse all of the parsing logic for each of the args
5. I went with raw string literals for the usage blocks and tweaked that output
6. I improved the handling of the token fallback and usage output when neither the arg nor env var is specified

I also remembered that in the review meeting, we determined we should force the label prefix to end in a non-letter, non-number. I added that logic.

Sorry for the unexpected churn and needing 1 more review. Thanks though!

RussKie commented on src/Downloader/Args.cs:
```diff
@@ -0,0 +1,226 @@
+                --page-limit        Maximum number of pages to retrieve.
+                --retries           Comma-separated retry delays in seconds. Default: 30,30,300,300,3000,3000.
+                --token             GitHub access token. Default: Read from GITHUB_TOKEN env var.
+                --verbose           Enable verbose output.
```
[nit] GHC was giving me a similar suggestion, but I found it rather unclear - that was the reason to list several usage patterns (where applicable).
This is not a blocking comment, but something to consider nevertheless.

In the specific case of the Downloader the required args are only `--repo {{org/repo1}}[,{{org/repo2}},...] --label-prefix {{label-prefix}}`. All other args are optional.

jeffhandley replied:
They're _sort of_ optional. You have to specify at least one of `--issue-data` or `--pull-data`. But there's no perfect way to express it. And... good enough is good enough.

RussKie commented:
> @RussKie I did more testing and decided to make another round of changes to the arg parsing.
> 
> 1. Changed the names from 'Config' to 'Args' and collapsed from 2 files into 1, putting the parse functions into the structs as statics (instead of having to bounce between the 2 files)
> 2. I found the issue-data and pull-data default value handling wasn't working as I thought it was. Those defaults didn't allow choosing to only process one or the other and it applied the default to make it always run both. We need to be able to run them independently.
> 3. I matched up a couple of arg struct property names that weren't consistent (IssuesPath vs. IssueDataPath)
> 4. I refactored out some arg helpers to reuse all of the parsing logic for each of the args
> 5. I went with raw string literals for the usage blocks and tweaked that output
> 6. I improved the handling of the token fallback and usage output when neither the arg nor env var is specified
> 
> I also remembered that in the review meeting, we determined we should force the label prefix to end in a non-letter, non-number. I added that logic.
> 
> Sorry for the unexpected churn and needing 1 more review. Thanks though!

👍

RussKie commented on .github/workflows/labeler-cache-retention.yml:
```diff
@@ -0,0 +1,13 @@
+
+jobs:
+  cache-retention:
+    # Do not run the workflow on forks outside the 'dotnet' org
+    if: ${{ github.repository_owner == 'dotnet' }}
+    uses: dotnet/issue-labeler/.github/workflows/cache-retention.yml@main
```
How would an external consumer exercise cache retention? E.g., the labeller is used in a microsoft/repo.

jeffhandley replied:
There are 2 sets of workflows in place in this repo:

1. The _consuming_ workflows that each repository that onboards will have in place, invoked through `workflow_dispatch` (manual) and other events
2. The _reusable_ workflows in this repository that implement the processes and can be invoked via `workflow_call`

The _consuming_ workflows call the _resuable_workflows as though they are actions, with the `uses: dotnet/issue-labeler/**/*.yml`. These the workflows that implement the "only for `dotnet`" logic, as these are the repos that would be forked and the jobs should be skipped. The _reusable_ workflows do not have those kinds of restriction logic.

In this repo and this repo only, we have both sets of worklows. The ones prefixed with `labeler-` are the _consuming_ workflows, serving as the examples that other repositories will copy to onboard while also ensuring this repository is testing itself.

Long story short, the consuming repos can set up the `repository_owner` restrictions as they see fit. We will restrict to `dotnet` for all our dotnet org repos, and we'd restrict to `microsoft` for those repos. There might also be scenarios where the repo will want to lock down to the specific owner/repo name so that forks within the same org also skip the jobs.

RussKie commented on src/Downloader/Args.cs:
```diff
@@ -0,0 +1,180 @@
+        Queue<string> arguments = new(args);
+        while (arguments.Count > 0)
+        {
+            string argument = arguments.Dequeue();
```
I'm curious whether you considered using System.CommandLine functionality, and if yes - what were the reason to discard it?

jeffhandley replied:
Definitely did. And I do expect to adopt S.CL at some point. We've been working on a pending new version of S.CL and I want to wait for it to be ready, and then that'll also give us a baseline of whether adopting it affects the live labeling speed/cost we observe after rollout.
